src/NovelRT/DotNet/RuntimeService.cpp

```diff
@@ -0,0 +1,148 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root for more information.
+
+#include <NovelRT.h>
+
+#include <nethost.h>
+
+#if defined(WIN32)
+#include <Windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+using namespace NovelRT::Utilities;
+
+#if defined(WIN32)
+  #define STR(s) L ## s
+  #define CH(c) L ## c
+#else
+  #define STR(s) s
+  #define CH(c) c
+#endif
+
+const auto HostApiBufferTooSmall = 0x80008098;
+
+static void* getNativeExport(void* nativeLibrary, const char* exportName) {
+#if defined(_WIN32)
+  void* nativeExport = GetProcAddress(static_cast<HMODULE>(nativeLibrary), exportName);
+  assert(nativeExport != nullptr);
+  return nativeExport;
+#else
+  void* nativeExport = dlsym(nativeLibrary, exportName);
+  assert(nativeExport != nullptr);
+  return nativeExport;
+#endif
+}
+
+static void* loadNativeLibrary(const char_t* path)
+{
+#if defined(_WIN32)
+  HMODULE nativeLibrary = LoadLibraryW(path);
+  assert(nativeLibrary != nullptr);
+  return static_cast<void*>(nativeLibrary);
+#else
+  void* nativeLibrary = dlopen(path, RTLD_LAZY | RTLD_LOCAL);
+  assert(nativeLibrary != nullptr);
+  return nativeLibrary;
+#endif
+}
+
+static void closeNativeLibrary(void* nativeLibrary)
+{
+#if defined(_WIN32)
+  BOOL result = FreeLibrary(static_cast<HMODULE>(nativeLibrary));
+  assert(result != FALSE);
+#else
+  int result = dlclose(nativeLibrary);
+  assert(result == 0);
+#endif
+}
+
+static void* loadHostFxr() {
+  size_t buffer_size;
+  int result = get_hostfxr_path(nullptr, &buffer_size, nullptr);
+
+  if (result != HostApiBufferTooSmall)
+  {
+    std::cerr << "ERROR: failed to locate hostfxr" << std::endl;
+    throw - 1;
+  }
+
+  auto buffer = std::vector<char_t>(buffer_size);
+  result = get_hostfxr_path(buffer.data(), &buffer_size, nullptr);
+
+  if (result != 0)
+  {
+    std::cerr << "ERROR: failed to locate hostfxr" << std::endl;
+    throw - 1;
```

Comment:
We should probably throw `std::runtime_error` here in the same vein as the other places (even if its not what we want to do in the future, it'd be nice to have this uniform), unless this is going to be exported as a C function or something?

Reply:
Ah right, forgot to fix these now that we have a logger.